### PR TITLE
stress_schedule: Fix for !GIL + emit=native.

### PR DESCRIPTION
**Pull request by jepler**
### Summary

Ensure that zero-duration sleeps run scheduled callbacks.

Reduce a wait making the test run 8x faster locally, which hopefully also makes it reliable on ci.

Restore the test during ci as well as some asyncio tests.

Closes: #18415 

ping @projectgus 

### Testing

I ran the script manually locally.

### Trade-offs and Alternatives

We could delete stress_schedule, it's been unused in CI since 2020.

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks, but I'm not sure this is the correct fix.  Instead, I think `time.sleep(0)` should be guaranteed to run the scheduler (eg via `mp_handle_pending()`).  Then this test would be fine as-is.

(So, this test is revealing a bug in the unix port, that `time.sleep(0)` and `time.sleep_ms(0)` exit without checking for pending events.)

**Comment by jepler** [comment]
I can look at it from that angle instead, thanks.

File: extmod/modtime.c [hunk 1]
```diff
@@ -134,6 +134,8 @@ static mp_obj_t time_sleep_ms(mp_obj_t arg) {
     mp_int_t ms = mp_obj_get_int(arg);
     if (ms >= 0) {
         mp_hal_delay_ms(ms);
+    } else {
+        mp_handle_pending(true);
```
Comment (dpgeorge) [inline extmod/modtime.c]: Hmm, I'm not sure this bit needs to be added.  There doesn't seem to be anything broken that needs this change.

Comment (jepler) [reply extmod/modtime.c]: It's not needed to pass the test as-written, but it's needed to pass this variant (with `-X emit=native`):
```diff
 while n < _NUM_TASKS and time.ticks_diff(time.ticks_ms(), t) < _TIMEOUT_MS:
-    time.sleep(0)
+    time.sleep_ms(0)
```
I can revert this part if you prefer, and document that `sleep(0)` is defined to run pending callbacks while `sleep_ms(0)` does not.

I investigated unconditionally calling `mp_hal_delay_ms` here instead but at least on Unix doing so does not handle pending callbacks and I didn't go deeper.

[accidentally posted as a draft review yesterday, belatedly actually publising today!]

Comment (dpgeorge) [reply extmod/modtime.c]: I don't understand.

The existing logic here is `if (ms >= 0)`, so this change won't change anything for `time.sleep_ms(0)`.

Comment (jepler) [reply extmod/modtime.c]: You're right, I must have tested incorrectly with sleep ms.

Comment (jepler) [reply extmod/modtime.c]: reverted, & added a change down in unix_mphal.c instead.

the `sleep_ms(0)` case is not explicitly tested but I tested it locally.

File: ports/unix/modtime.c [hunk 2]
```diff
@@ -95,6 +95,7 @@ static mp_obj_t mp_time_sleep(mp_obj_t arg) {
     tv.tv_sec = (suseconds_t)ipart;
     int res;
     while (1) {
+        mp_handle_pending(true);
```
Comment (dpgeorge) [inline ports/unix/modtime.c:98]: Need to think if it's best to have this call here, or just after the `MP_THREAD_GIL_ENTER()` below.

The point is, events are likely to happen during a sleep, and then it's minimal latency to process such events as soon as the sleep is finished.

OTOH, maybe there are events already queued, and they should definitely be processed before sleeping.

Note that esp32 implements it how you have it here.

File: ports/unix/unix_mphal.c [hunk 3]
```diff
@@ -243,6 +243,7 @@ uint64_t mp_hal_time_ns(void) {
 #ifndef mp_hal_delay_ms
 void mp_hal_delay_ms(mp_uint_t ms) {
     mp_uint_t start = mp_hal_ticks_ms();
+    mp_handle_pending(true);
```
Comment (dpgeorge) [inline ports/unix/unix_mphal.c]: There's already a call to `mp_handle_pending(true)` at the start of `mp_event_wait_ms()`.  So that's a bit inefficient to call it twice in a row.

Could follow how rp2 works:
```c
    mp_uint_t start = mp_hal_ticks_ms(); 
    mp_uint_t elapsed = 0;
    do {                                                           
        mp_event_wait_ms(ms - elapsed);                                     
        elapsed = mp_hal_ticks_ms() - start;                        
    } while (elapsed < ms);              
```
But that would need modifying `MICROPY_INTERNAL_WFE` to not sleep if the `TIMEOUT_MS` argument is 0.

Comment (jepler) [reply ports/unix/unix_mphal.c]: what do you think about `if (ticks_ms) { loop on mp_event_wait } else { mp_handle_pending(true); }`?

Do we need an issue filed to examine or test all implementations of `mp_hal_wait_ms` and `mp_time_sleep` for the zero wait case?

Comment (dpgeorge) [reply ports/unix/unix_mphal.c]: > what do you think about `if (ticks_ms) { loop on mp_event_wait } else { mp_handle_pending(true); }`?

Yes, that sounds fine.



> Do we need an issue filed to examine or test all implementations of `mp_hal_wait_ms` and `mp_time_sleep` for the zero wait case?

Yes, probably worth adding a test.  There's already `tests/micropython/schedul.py` which does some testing.  Although it's excluded when built with the native emitter because it does busy loops, and the native emitter won't run the scheduler from a busy loop.  So, probably "copying" that test, changing the busy loops to `sleep(0)` and `sleep_ms(0)` would be enough.

**Comment by jepler** [comment]
I changed unix sleep_ms as requested and added to the documentation the guarantee (possibly not yet met in all ports!) that sleep(0) and sleep_ms(0) run scheduled tasks. I also found that some asyncio tests would run under the native emitter provided that `poll_poll_internal` C function called pending tasks.

Local testing showed that qemu's sleep_ms needed the added code to run callbacks for sleep_ms(0). Thus, it's likely that ALL ports should have their implementations checked & specifically the micropython/schedule.py test run under the native emitter.

 I'll check back and see how CI is doing.

File: ports/unix/unix_mphal.c (hunk 3 above)
Comment (jepler) [reply ports/unix/unix_mphal.c]: I modified the existing test and made it run under the native emitter; should I instead duplicate it?

Comment (dpgeorge) [reply ports/unix/unix_mphal.c]: I think it would be better to duplicate it, because the existing one is testing that the VM calls the scheduler without explicit calls to sleep/poll/etc.

Comment (jepler) [reply ports/unix/unix_mphal.c]: done.

File: extmod/modselect.c [hunk 4]
```diff
@@ -537,6 +537,8 @@ MP_DEFINE_CONST_FUN_OBJ_3(poll_modify_obj, poll_modify);
 static mp_uint_t poll_poll_internal(uint n_args, const mp_obj_t *args) {
     mp_obj_poll_t *self = MP_OBJ_TO_PTR(args[0]);
 
+    mp_handle_pending(true);
```
Comment (dpgeorge) [inline extmod/modselect.c]: I suggest moving this to the start of `poll_set_poll_until_ready_or_timeout()`, so that it's also called for `select.select()`.

File: ports/unix/modtime.c (hunk 2 above)
Comment (dpgeorge) [reply ports/unix/modtime.c:98]: This looks fine to me.

File: tests/micropython/schedule_sleep.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
+# test micropython.schedule() function
+
```
Comment (dpgeorge) [inline tests/micropython/schedule_sleep.py:5]: Suggest adding a comment here:
```py
# this is the same as tests/micropython/schedule.py but the busy loops are replaced with sleep/sleep_ms
```

File: extmod/modselect.c (hunk 4 above)
Comment (jepler) [reply extmod/modselect.c]: done.

File: tests/micropython/schedule_sleep.py (hunk 5 above)
Comment (jepler) [reply tests/micropython/schedule_sleep.py:5]: done

**Comment by dpgeorge** [comment]
Thanks for updating.

The CI failures look unrelated to this PR's changes.  I'm guessing the Python version changed in the Windows CI and that led to settrace results differing?  Well, that will need to be investigated separately.

**Comment by jepler** [comment]
I double checked that my branch was up to date with master, and it is. (though I think CI tests the merge commit anyway? not sure)

**Review by dpgeorge — APPROVED** [review]
This looks good now, thanks.

I also tested it on RP_PICO_W and ESP32_GENERIC_C3, running the tests in native mode, and they pass.

**Comment by dpgeorge** [comment]
Rebased and merged in b792efc341be16da31e1b5d855e61307e1047310 through f918e336452b8506ca30887fc37e92733718e3bb
